### ISSUE TITLE
[QA] Windows11使用bitsandbytes运行InternLM2-chat-7B-4bits量化，大模型精神错乱

### ISSUE THREAD

**Issue by Scoodtone** [InternLM/InternLM#680]
### Describe the question.

![image](https://github.com/InternLM/InternLM/assets/76857360/7730756a-7fb9-40b1-952b-49c2606002c1)
![image](https://github.com/InternLM/InternLM/assets/76857360/7c927988-70fd-4f79-8f0f-e31e614cbfe4)
如图，大模型会无视用户输入并输出一些疑似知识库内的问答内容，且输出一遍问答内容之后会重复之前输出的内容持续输出到自动终止

**Comment by Scoodtone** [comment]
Python 3.10.13

**Comment by Scoodtone** [comment]
![image](https://github.com/InternLM/InternLM/assets/76857360/7ee2db0d-de8e-4a62-9300-3724b6895615)

**Comment by Scoodtone** [comment]
![image](https://github.com/InternLM/InternLM/assets/76857360/f0a200bc-6fc7-4c57-83cf-3a6344a784ef)

**Comment by lvhan028** [comment]
internlm2-chat-7b-4bits是用lmdeploy量化的，可以直接用lmdeploy推理。没有验证过它是否能直接对接bitsandbytes。

**Comment by ZwwWayne** [comment]
看起来是没有停下来，请问下transformer版本是？

**Comment by Scoodtone** [comment]
> 看起来是没有停下来，请问下transformer版本是？

4.37.2，麻烦您了

**Comment by ZwwWayne** [comment]
这个的原因应该是你设置的 max_length 最小长度为 32，但是模型普通的回复达不到 32 的长度，所以导致模型不会停下。可以在 web_demo 的脚本中 https://github.com/InternLM/InternLM/blob/main/chat/web_demo.py#L194 修改调小最小值


在相同transformers 版本下，load_in_4bit，cli 脚本中不会遇到这个问题。示例代码如下：
```
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

model_path = "internlm/internlm2-chat-7b/"
print('initializing model')
model = AutoModelForCausalLM.from_pretrained(model_path, device_map='auto', trust_remote_code=True, load_in_4bit=True)
print('initialized model')
tokenizer = AutoTokenizer.from_pretrained(model_path, trust_remote_code=True)
print('initialized tokenizer')
model = model.eval()
length = 0
response, history = model.chat(tokenizer, "你是谁", history=[])
print(response[0:], flush=True, end="\n")
length += len(response)
response, history = model.chat(tokenizer, "你好", history=history)
print(response[0:], flush=True, end="\n")
print(history)
```
输出结果如下：
<img width="1070" alt="image" src="https://github.com/InternLM/InternLM/assets/40779233/034e14f5-7579-41cd-ab75-c55e7568b4aa">

**Comment by Scoodtone** [comment]
明白了，但是max_length不应该是模型回复的最大长度吗，我翻了一下Transformers的interface.py，发现好像没有min_length这个选项(附图一)，另外我尝试了回答大于32的提问，最终发现他会正常回答问题，但本该停止的时候他会继续回答永不停止，怀疑可能和issues其他人遇到的那个不停止的问题很像(附图2)，麻烦您了！
![image](https://github.com/InternLM/InternLM/assets/76857360/6950f2e0-cd44-4095-85de-44c3259cb83a)

![image](https://github.com/InternLM/InternLM/assets/76857360/57daeef4-b0d8-472c-9864-511f51287bdf)

**Comment by ZwwWayne** [comment]
可能和一些对话超参数也有关系，主要是之前的一个 case 命令行没有复现，有点奇怪
也可以检查下HF的代码是否完整更新

**Comment by ZwwWayne** [comment]
<img width="1306" alt="image" src="https://github.com/InternLM/InternLM/assets/40779233/d2bc7dc4-affe-47ec-9c0f-60b305c6f591">
用 https://github.com/InternLM/InternLM/blob/main/chat/web_demo.py 似乎没有复现，可以看看用这个项目中最新的脚本会不会遇到

**Comment by Scoodtone** [comment]
嗯嗯可能是参数的问题，更换新demo后问题没有出现了，谢谢佬！另外提下新demo里avatar的存放路径和旧的不一样，以后有人问的话可以提醒一下~(顺便还想问一下**Lagent+InternLM2能否实现读取链接的功能**，最近想做一个项目，项目要求大模型能够浏览页面内容(如csdn)，如果可以的话我继续进一步做)(其实还想实现读图，这个我知道咱双卡Xcomposer可以做到，但电脑带不起来，等之后读取链接成功实现之后用咱Studio云算力试一下Xcomposer)

**Comment by ZwwWayne** [comment]
直接浏览网页我们还没尝试过，读图的话建议尝试 xcomposer

**Comment by Scoodtone** [comment]
> 直接浏览网页我们还没尝试过，读图的话建议尝试 xcomposer

好的，那我之后使用Lagent尝试一下，谢谢！

**Comment by XARKUR** [comment]
我使用langchain-chatchat跑也会遇到类似的问题，就前一两个问题的回答是正常的，后面就开始循环了，不知道怎么解决

**Comment by Scoodtone** [comment]
我又来了佬，我把streamlit删了改成Flask了，准备用POST跟Node.js的后端连一下，用Postman测试了一下发现又出现复读机了，代码是基于上次没出错的新web_demo.py修改的，很多地方不懂都没敢改，具体附下：

**Comment by Scoodtone** [comment]
`
# isort: skip_file
import copy
import warnings
from dataclasses import asdict, dataclass
from typing import Callable, List, Optional

import torch
from torch import nn
from transformers.generation.utils import (LogitsProcessorList,
                                           StoppingCriteriaList)
from transformers.utils import logging

from transformers import AutoTokenizer, AutoModelForCausalLM  # isort: skip

from flask import Flask, request, jsonify
app = Flask(__name__)

logger = logging.get_logger(__name__)

@dataclass
class GenerationConfig:
    # this config is used for chat to provide more diversity
    max_length: int = 32768
    top_p: float = 0.8
    temperature: float = 0.8
    do_sample: bool = True
    repetition_penalty: float = 1.005

@torch.inference_mode()
def generate_interactive(
    model,
    tokenizer,
    prompt,
    generation_config: Optional[GenerationConfig] = None,
    logits_processor: Optional[LogitsProcessorList] = None,
    stopping_criteria: Optional[StoppingCriteriaList] = None,
    prefix_allowed_tokens_fn: Optional[Callable[[int, torch.Tensor],
                                                List[int]]] = None,
    additional_eos_token_id: Optional[int] = None,
    **kwargs,
):
    inputs = tokenizer([prompt], padding=True, return_tensors='pt')
    input_length = len(inputs['input_ids'][0])
    for k, v in inputs.items():
        inputs[k] = v.cuda()
    input_ids = inputs['input_ids']
    _, input_ids_seq_length = input_ids.shape[0], input_ids.shape[-1]
    if generation_config is None:
        generation_config = model.generation_config
    generation_config = copy.deepcopy(generation_config)
    model_kwargs = generation_config.update(**kwargs)
    bos_token_id, eos_token_id = (  # noqa: F841  # pylint: disable=W0612
        generation_config.bos_token_id,
        generation_config.eos_token_id,
    )
    if isinstance(eos_token_id, int):
        eos_token_id = [eos_token_id]
    if additional_eos_token_id is not None:
        eos_token_id.append(additional_eos_token_id)
    has_default_max_length = kwargs.get(
        'max_length') is None and generation_config.max_length is not None
    if has_default_max_length and generation_config.max_new_tokens is None:
        warnings.warn(
            f"Using 'max_length''s default ({repr(generation_config.max_length)}) \
                to control the generation length. "
            'This behaviour is deprecated and will be removed from the \
                config in v5 of Transformers -- we'
            ' recommend using `max_new_tokens` to control the maximum \
                length of the generation.',
            UserWarning,
        )
    elif generation_config.max_new_tokens is not None:
        generation_config.max_length = generation_config.max_new_tokens + \
            input_ids_seq_length
        if not has_default_max_length:
            logger.warn(  # pylint: disable=W4902
                f"Both 'max_new_tokens' (={generation_config.max_new_tokens}) "
                f"and 'max_length'(={generation_config.max_length}) seem to "
                "have been set. 'max_new_tokens' will take precedence. "
                'Please refer to the documentation for more information. '
                '(https://huggingface.co/docs/transformers/main/'
                'en/main_classes/text_generation)',
                UserWarning,
            )

    if input_ids_seq_length >= generation_config.max_length:
        input_ids_string = 'input_ids'
        logger.warning(
            f"Input length of {input_ids_string} is {input_ids_seq_length}, "
            f"but 'max_length' is set to {generation_config.max_length}. "
            'This can lead to unexpected behavior. You should consider'
            " increasing 'max_new_tokens'.")

    # 2. Set generation parameters if not already defined
    logits_processor = logits_processor if logits_processor is not None \
        else LogitsProcessorList()
    stopping_criteria = stopping_criteria if stopping_criteria is not None \
        else StoppingCriteriaList()

    logits_processor = model._get_logits_processor(
        generation_config=generation_config,
        input_ids_seq_length=input_ids_seq_length,
        encoder_input_ids=input_ids,
        prefix_allowed_tokens_fn=prefix_allowed_tokens_fn,
        logits_processor=logits_processor,
    )

    stopping_criteria = model._get_stopping_criteria(
        generation_config=generation_config,
        stopping_criteria=stopping_criteria)
    logits_warper = model._get_logits_warper(generation_config)

    unfinished_sequences = input_ids.new(input_ids.shape[0]).fill_(1)
    scores = None
    while True:
        model_inputs = model.prepare_inputs_for_generation(
            input_ids, **model_kwargs)
        # forward pass to get next token
        outputs = model(
            **model_inputs,
            return_dict=True,
            output_attentions=False,
            output_hidden_states=False,
        )

        next_token_logits = outputs.logits[:, -1, :]

        # pre-process distribution
        next_token_scores = logits_processor(input_ids, next_token_logits)
        next_token_scores = logits_warper(input_ids, next_token_scores)

        # sample
        probs = nn.functional.softmax(next_token_scores, dim=-1)
        if generation_config.do_sample:
            next_tokens = torch.multinomial(probs, num_samples=1).squeeze(1)
        else:
            next_tokens = torch.argmax(probs, dim=-1)

        # update generated ids, model inputs, and length for next step
        input_ids = torch.cat([input_ids, next_tokens[:, None]], dim=-1)
        model_kwargs = model._update_model_kwargs_for_generation(
            outputs, model_kwargs, is_encoder_decoder=False)
        unfinished_sequences = unfinished_sequences.mul(
            (min(next_tokens != i for i in eos_token_id)).long())

        output_token_ids = input_ids[0].cpu().tolist()
        output_token_ids = output_token_ids[input_length:]
        for each_eos_token_id in eos_token_id:
            if output_token_ids[-1] == each_eos_token_id:
                output_token_ids = output_token_ids[:-1]
        response = tokenizer.decode(output_token_ids)

        yield response
        # stop when each sentence is finished
        # or if we exceed the maximum length
        if unfinished_sequences.max() == 0 or stopping_criteria(
                input_ids, scores):
            break

def load_model():
    model = (AutoModelForCausalLM.from_pretrained('D:/InternLM/InternLM2-chat-7B',
                                                  trust_remote_code=True, load_in_4bit=True)
                                                  
                                                  #.to(torch.bfloat16).cuda()
                                                  )
    tokenizer = AutoTokenizer.from_pretrained('D:/InternLM/InternLM2-chat-7B',
                                              trust_remote_code=True)
    return model, tokenizer

def prepare_generation_config():
    generation_config = GenerationConfig(max_length=32768,
                                         top_p=0.8,
                                         temperature=0.8,
                                         do_sample=True,
                                         repetition_penalty=1.005)
    return generation_config

user_prompt = '<|im_start|>user\n{user}<|im_end|>\n'
robot_prompt = '<|im_start|>assistant\n{robot}<|im_end|>\n'
cur_query_prompt = '<|im_start|>user\n{user}<|im_end|>\n\
    <|im_start|>assistant\n'

def combine_history(prompt, history):
    #messages = st.session_state.messages
    meta_instruction = ('You are InternLM (书生·浦语), a helpful, honest, '
                        'and harmless AI assistant developed by Shanghai '
                        'AI Laboratory (上海人工智能实验室).')
    total_prompt = f"<s><|im_start|>system\n{meta_instruction}<|im_end|>\n"
    for message in history:
        cur_content = message['content']
        if message['role'] == 'user':
            cur_prompt = user_prompt.format(user=cur_content)
        elif message['role'] == 'robot':
            cur_prompt = robot_prompt.format(robot=cur_content)
        else:
            raise RuntimeError
        total_prompt += cur_prompt
    total_prompt = total_prompt + cur_query_prompt.format(user=prompt)
    return total_prompt

print('load model begin.')
model, tokenizer = load_model()
print('load model end.')        

@app.route('/analyze', methods=['POST'])
def analyze_text():

    # torch.cuda.empty_cache()
    
    generation_config = prepare_generation_config()

    data = request.json
    prompt = data['text']
    history = data['history']  # 假设历史消息以列表形式发送
    combined_prompt = combine_history(prompt, history)
    result = generate_interactive(model=model,
                    tokenizer=tokenizer,
                    prompt=combined_prompt,
                    additional_eos_token_id=92542,
                    **asdict(generation_config),)
    # 收集生成器的所有输出到一个字符串
    full_response = ''.join(list(result))

    return jsonify({"result": full_response})

if __name__ == '__main__':
    app.run(debug=True, port=6006)

`

**Comment by Scoodtone** [comment]
代码如下，上传到个人主页方便查看：https://github.com/Scoodtone/InternLM_Try/blob/main/web_demo_razor.py

**Comment by Scoodtone** [comment]
Postman发送的内容与接收到的回答如下：
![image](https://github.com/InternLM/InternLM/assets/76857360/1b4a1291-5c0c-4024-a37c-fc55e0c25da4)
